CRM/Core/Menu.php

```diff
@@ -265,14 +265,19 @@ public static function fillMenuValues(&$menu, $path) {
    */
   public static function build(&$menu) {
     foreach ($menu as $path => $menuItems) {
-      self::buildBreadcrumb($menu, $path);
-      self::fillMenuValues($menu, $path);
-      self::fillComponentIds($menu, $path);
-      self::buildReturnUrl($menu, $path);
-
-      // add add page_type if not present
-      if (!isset($menu[$path]['page_type'])) {
-        $menu[$path]['page_type'] = 0;
+      try {
+        self::buildBreadcrumb($menu, $path);
+        self::fillMenuValues($menu, $path);
+        self::fillComponentIds($menu, $path);
+        self::buildReturnUrl($menu, $path);
+
+        // add add page_type if not present
+        if (!isset($menu[$path]['page_type'])) {
+          $menu[$path]['page_type'] = 0;
+        }
+      }
+      catch (CRM_Core_Exception $e) {
+        CRM_Core_Session::setStatus($e->getMessage(), ts('Path skipped'));
```

Comment:
In general I think `CRM_Core_Session::setStatus()` is only meant for use on forms, and lower-down BAO and Util classes should *not* call it. Consider the case of an API calling a BAO function and that BAO function calling `CRM_Core_Session::setStatus()` to handle errors. In that case the api might even be running over REST or other headless state so there isn't even a user session to set messages for.

Reply:
yeah -  but then it just gets swallowed doesn't it? Which seems OK - outside forms.

Reply:
Well it's led to some weird bugs in the past. Typically we expect the BAO to throw an exception and the layer that calls it to catch it and respond appropriately (if it's an api, generate an api error, if it's a form, display a status message, etc).

Reply:
The problem is that the iterate-through-array-of-menu-items is still at the BAO layer - we want that layer to continue. I guess it could throw an exception at the end of the iteration - but then it might still kill api calls

Reply:
@eileenmcnaughton I recommend logging the error instead of calling `setStatus()`

Reply:
OK - I've updated it

Reply:
I don't know everything about logging in Civi but wouldn't this be classified as an error message rather than a debug message? e.g. `Civi::log()->error()`

Reply:
@colemanw done - can we merge now?